File: Snake_Java/src/mathengine/GameMath.java
package mathengine;

import java.util.Random;

/**
 * GameMath.java
 *
 * Contains methods to calculate the x and y
 * bounds for randomising the coordinates
 * to place the food and poisons.
 */
public class GameMath
{
    /**
     * Generate a random x position for left screen
     *
     * @param rand:     Random object
     * @param center:   Center screen position
     * @param unitSize: Size of unit
     * @return randomly selected x position for left screen
     */
    public static int getRandomLeftX(Random rand, int center, int unitSize)
    {
        return rand.nextInt(calculateXBound(center, unitSize)) * unitSize;
    }

    /**
     * Generate a random x position for right screen
     *
     * @param rand:     Random object
     * @param center:   Center screen position
     * @param unitSize: Size of unit
     * @return randomly selected x position for right screen
     */
    public static int getRandomRightX(Random rand, int center, int unitSize)
    {
        return ((rand.nextInt(calculateXBound(center, unitSize))) + calculateXBound(center, unitSize)) * unitSize;
    }

    /**
     * Generate a random y position for game screen
     *
     * @param rand:     Random object
     * @param yPos:     y position
     * @param height:   Height of game window
     * @param unitSize: Size of unit
     * @return randomly generated y position
     */
    public static int getRandomY(Random rand, int yPos, int height, int unitSize)
    {
        return yPos + (rand.nextInt(GameMath.calculateYBound(height, unitSize)) * unitSize);
    }

    /**
     * Calculate bound for random generating of x position
     *
     * @param center:   Center x position
     * @param unitSize: Size of unit
     * @return bound
     */
    private static int calculateXBound(int center, int unitSize)
    {
        return center / unitSize;
    }

    /**
     * Calculate bound for random generating of y position
     *
     * @param height:   Height of window
     * @param unitSize: Size of unit
     * @return bound
     */
    private static int calculateYBound(int height, int unitSize)
    {
        return (height-unitSize) / unitSize;
    }
}

File: Snake_C++/src/Math/GameMath.cpp
#include "../../include/GameMath.hpp"
#include <cstdlib>
#include <iostream>
#include <exception>
//#include <windows.h>

using namespace Math;


struct MultiplyByZero : public std::exception {
    const char * what () const throw () {
        return "MultiplyByZero Exception";
    }
};


struct DivideByZero : public std::exception {
    const char * what () const throw () {
        return "DivideByZero Exception";
    }
};


/**
 * Calculate bound for random generating of x position
 *
 * @param center:   Center x position
 * @param unitSize: Size of unit
 * @return bound
 */
int GameMath::calculateXBound(int center, int unitSize)
{
    try {
        if (unitSize == 0) {
            throw DivideByZero();
        }
    } catch(DivideByZero& e) {
        std::cerr << "DivideByZero exception caught at GameMath::calculateXBound()! unitSize == 0" << std::endl;
        std::cerr << e.what() << std::endl;
//        MessageBox(NULL, "DivideByZero error!",
//                   "Exception has occurred!", MB_ICONERROR); // windows only.
        std::terminate(); // aborts program.
    }
    return center / unitSize;
}

/**
 * Calculate bound for random generating of y position
 *
 * @param height:   Height of window
 * @param unitSize: Size of unit
 * @return bound
 */
int GameMath::calculateYBound(int height, int unitSize)
{
    try {
        if (unitSize == 0) {
            throw DivideByZero();
        }
    } catch(DivideByZero& e) {
        std::cerr << "DivideByZero exception caught at GameMath::calculateYBound()! unitSize == 0" << std::endl;
        std::cerr << e.what() << std::endl;
//        MessageBox(NULL, "DivideByZero error!",
//                   "Exception has occurred!", MB_ICONERROR); // windows only.
        std::terminate(); // aborts program.
    }
    return (height - unitSize) / unitSize;
}

/**
 * Generate a random y position for game screen
 *
 * @param yPos:     y position
 * @param height:   Height of game window
 * @param unitSize: Size of unit
 * @return randomly generated y position
 */
int GameMath::getRandomY(int yPos, int height, int unitSize)
{
    try {
        if (unitSize == 0) {
            throw MultiplyByZero();
        }
    } catch(MultiplyByZero& e) {
        std::cerr << "MultiplyByZero exception caught at GameMath::getRandomY()! unitSize == 0" << std::endl;
        std::cerr << e.what() << std::endl;
//        MessageBox(NULL, "MultiplyByZero error!",
//                   "Exception has occurred!", MB_ICONERROR); // windows only.
        std::terminate(); // aborts program.
    }
    return yPos + rand() % calculateYBound(height, unitSize) * unitSize;
}

/**
 * Generate a random x position for left screen
 *
 * @param center:   Center screen position
 * @param unitSize: Size of unit
 * @return randomly selected x position for left screen
 */
int GameMath::getRandomLeftX(int center, int unitSize)
{
    try {
        if (unitSize == 0) {
            throw MultiplyByZero();
        }
    } catch(MultiplyByZero& e) {
        std::cerr << "MultiplyByZero exception caught at GameMath::getRandomLeftX()! unitSize == 0" << std::endl;
        std::cerr << e.what() << std::endl;
//        MessageBox(NULL, "MultiplyByZero error!",
//                   "Exception has occurred!", MB_ICONERROR); // windows only.
        std::terminate(); // aborts program.
    }
    return rand() % calculateXBound(center, unitSize) * unitSize;
}

/**
 * Generate a random x position for right screen
 *
 * @param center:   Center screen position
 * @param unitSize: Size of unit
 * @return randomly selected x position for right screen
 */
int GameMath::getRandomRightX(int center, int unitSize)
{
    try {
        if (unitSize == 0) {
            throw MultiplyByZero();
        }
    } catch(MultiplyByZero& e) {
        std::cerr << "MultiplyByZero exception caught at GameMath::getRandomRightX()! unitSize == 0" << std::endl;
        std::cerr << e.what() << std::endl;
//        MessageBox(NULL, "MultiplyByZero error!",
//                   "Exception has occurred!", MB_ICONERROR); // windows only.
        std::terminate(); // aborts program.
    }
    return (rand() % calculateXBound(center, unitSize) + calculateXBound(center, unitSize)) * unitSize;
}
